onnxruntime/core/graph/graph.cc

```diff
@@ -2055,6 +2056,39 @@ void Graph::RemoveInitializedTensor(const std::string& tensor_name) {
   }
 }
 
+Status Graph::ReplaceInitializedTensor(const ONNX_NAMESPACE::TensorProto& new_initializer) {
+  const auto& initializer_name = new_initializer.name();
+  auto name_to_initializer_it = name_to_initial_tensor_.find(initializer_name);
+  ORT_RETURN_IF_NOT(name_to_initializer_it != name_to_initial_tensor_.end(),
+                    "Failed to find existing initializer with name ", initializer_name, ".");
+
+  assert(name_to_initializer_it->second);
+  const auto& old_initializer = *(name_to_initializer_it->second);
+
+  auto dims_eq = [&old_initializer, &new_initializer]() {
+    if (old_initializer.dims_size() != new_initializer.dims_size()) return false;
+    for (int i = 0; i < old_initializer.dims_size(); ++i) {
+      if (old_initializer.dims(i) != new_initializer.dims(i)) return false;
+    }
+    return true;
+  };
+
+  ORT_RETURN_IF_NOT(dims_eq(), "Replacement tensor's dimensions do not match.");
```

Comment:
>dims_eq [](start = 20, length = 7)

I think this should be made optional, in case of changing weight layout, i.e. conv filters to NCHWc. When layout changes, the actual dimensions might change. Maybe add a boolean in the function arguments?

Reply:
An alternative is to call Graph::RemoveInitializedTensor() and Graph::AddInitializedTensor(). There's at least one known issue with that though (https://github.com/microsoft/onnxruntime/issues/2232).

Reply:
I would have thought if the layout was changing an initializer with a different name should be added (and the relevant nodes updated).